File: haskell/glue/src/Glue/Lib/List/Car.hs
module Glue.Lib.List.Car where

import Glue.Eval (Eval, throwError)
import Glue.Eval.Exception (wrongArgumentType)
import Glue.IR (IR (..))

car :: IR Eval
car = NativeFunc carImpl

carImpl :: IR Eval -> Eval (IR Eval)
carImpl arg = case arg of
    List (x : _) -> pure x
    List [] -> throwError $ wrongArgumentType ["non-empty list"]
    _ -> throwError $ wrongArgumentType ["list"]

File: dart/glue/lib/src/lib/list/car.dart
import 'package:glue/src/eval.dart';
import 'package:glue/src/eval/exception.dart';
import 'package:glue/src/ir.dart';

/// Car function - returns the first element of a list
/// Mirrors Haskell Glue.Lib.List.Car.car exactly
Ir car = IrNativeFunc(carImpl);

/// Car function implementation
/// Mirrors Haskell Glue.Lib.List.Car.carImpl exactly
Eval<Ir> carImpl(Ir arg) {
  return switch (arg) {
    IrList(:final elements) =>
      elements.isEmpty
          ? throwError(wrongArgumentType(['non-empty list']))
          : Eval.pure(elements.first),
    _ => throwError(wrongArgumentType(['list'])),
  };
}
